bpf/lib/fib.h

```diff
@@ -171,87 +171,19 @@ fib_do_redirect(struct __ctx_buff *ctx, const bool needs_l2_check,
 
 static __always_inline int
 fib_redirect(struct __ctx_buff *ctx, const bool needs_l2_check,
-	     struct bpf_fib_lookup_padded *fib_params, __s8 *fib_err, int *oif)
+	     struct bpf_fib_lookup_padded *fib_params __maybe_unused,
+	     __s8 *fib_err __maybe_unused, int *oif)
 {
-	struct bpf_redir_neigh nh_params;
-	struct bpf_redir_neigh *nh = NULL;
-	bool no_neigh = is_defined(ENABLE_SKIP_FIB);
-	int ret;
-
 #ifndef ENABLE_SKIP_FIB
+	int ret;
 	ret = fib_lookup(ctx, &fib_params->l, sizeof(fib_params->l), 0);
-	if (ret != BPF_FIB_LKUP_RET_SUCCESS) {
-		*fib_err = (__s8)ret;
-		if (likely(ret == BPF_FIB_LKUP_RET_NO_NEIGH)) {
-			nh_params.nh_family = fib_params->l.family;
-			__bpf_memcpy_builtin(&nh_params.ipv6_nh,
-					     &fib_params->l.ipv6_dst,
-					     sizeof(nh_params.ipv6_nh));
-			nh = &nh_params;
-			no_neigh = true;
-			/* For kernels without d1c362e1dd68 ("bpf: Always
-			 * return target ifindex in bpf_fib_lookup") we
-			 * fall back to use the caller-provided oif when
-			 * necessary.
-			 */
-			if (!is_defined(HAVE_FIB_IFINDEX) && *oif > 0)
-				goto skip_oif;
-		} else {
-			return DROP_NO_FIB;
-		}
-	}
-	*oif = fib_params->l.ifindex;
-skip_oif:
+	*fib_err = (__s8)ret;
+	return fib_do_redirect(ctx, needs_l2_check, fib_params, fib_err, oif);
 #else
+	__s8 skip_fib = BPF_FIB_LKUP_RET_NO_NEIGH;
 	*oif = DIRECT_ROUTING_DEV_IFINDEX;
-#endif /* ENABLE_SKIP_FIB */
-	if (needs_l2_check) {
-		bool l2_hdr_required = true;
-
-		ret = maybe_add_l2_hdr(ctx, *oif, &l2_hdr_required);
-		if (ret != 0)
-			return ret;
-		if (!l2_hdr_required)
-			goto out_send;
-	}
-	if (no_neigh) {
-		/* If we are able to resolve neighbors on demand, always
-		 * prefer that over the BPF neighbor map since the latter
-		 * might be less accurate in some asymmetric corner cases.
-		 */
-		if (neigh_resolver_available()) {
-			if (nh)
-				return redirect_neigh(*oif, &nh_params,
-						      sizeof(nh_params), 0);
-			else
-				return redirect_neigh(*oif, NULL, 0, 0);
-		} else {
-			union macaddr *dmac, smac = NATIVE_DEV_MAC_BY_IFINDEX(*oif);
-
-			/* The neigh_record_ip{4,6} locations are mainly from
-			 * inbound client traffic on the load-balancer where we
-			 * know that replies need to go back to them.
-			 */
-			dmac = fib_params->l.family == AF_INET ?
-			       neigh_lookup_ip4(&fib_params->l.ipv4_dst) :
-			       neigh_lookup_ip6((void *)&fib_params->l.ipv6_dst);
-			if (!dmac) {
-				*fib_err = BPF_FIB_MAP_NO_NEIGH;
-				return DROP_NO_FIB;
-			}
-			if (eth_store_daddr_aligned(ctx, dmac->addr, 0) < 0)
-				return DROP_WRITE_ERROR;
-			if (eth_store_saddr_aligned(ctx, smac.addr, 0) < 0)
-				return DROP_WRITE_ERROR;
-		}
-	} else {
-		if (eth_store_daddr(ctx, fib_params->l.dmac, 0) < 0)
-			return DROP_WRITE_ERROR;
-		if (eth_store_saddr(ctx, fib_params->l.smac, 0) < 0)
-			return DROP_WRITE_ERROR;
-	}
-out_send:
-	return ctx_redirect(ctx, *oif, 0);
+	return fib_do_redirect(ctx, needs_l2_check, NULL, &skip_fib, oif);
```

Comment:
When `fib_do_redirect()` sets an ext_err in its `fib_ret` parameter, it doesn't get propagated back to the caller from here. This means we lose the `BPF_FIB_MAP_NO_NEIGH` error info.

Reply:
Hmm, I think I had some feelings about this error message handling. 

For one, the error message is really for the `fib_lookup` error, at least semantically from the caller, and not about the redirect. 

I don't love that the error is kinda overloaded, leaving the application to kinda "guess" what actually went wrong. 

I think my goal here was to make the passed in `fib_err` actually represent the error returned from a fib_lookup, while the return value from the function to classify that a different type of error took place (in this case, a redirect error, or an error writing new TTL). 

What's your thoughts here? I'm also happy to just make error handling occur like it did previously, where the error variable was overloaded.

Reply:
Let's not mix subtle changes in behaviour into a "refactor" :). Or at least split them off into a separate patch that explains why this is better behaviour.

Reply:
Okay @julianwiedmann latest commit fixes  this issue. 

You are right, we still dirty the fib_ret error in the case where 'ENABLE_SKIP_FIB' macro is enabled. 

I think this is OK tho, as the caller should also check the return value of 'fib_do_redirect' and 'fib_redirect_v*' to determine what the issue was. 

The 'fib_ret' function now can pass the `BPF_FIB_MAP_NO_NEIGH` in case of error.